yarn-project/simulator/src/avm/avm_simulator.ts

```diff
@@ -79,13 +92,24 @@ export class AvmSimulator {
           'AVM attempted to execute non-existent instruction. This should never happen (invalid bytecode or AVM simulator bug)!',
         );
 
-        const gasLeft = `l2=${machineState.l2GasLeft} da=${machineState.daGasLeft}`;
-        this.log.debug(`@${machineState.pc} ${instruction.toString()} (${gasLeft})`);
+        const instrStartGas = machineState.gasLeft; // Save gas before executing instruction (for profiling)
+        const instrPc = machineState.pc; // Save PC before executing instruction (for profiling)
+
+        this.log.debug(
+          `@${machineState.pc} ${instruction.toString()} (l2=${machineState.l2GasLeft} da=${machineState.daGasLeft})`,
+        );
         // Execute the instruction.
         // Normal returns and reverts will return normally here.
         // "Exceptional halts" will throw.
         await instruction.execute(this.context);
 
+        // gas used by this instruction - used for profiling/tallying
+        const gasUsed: Gas = {
+          l2Gas: instrStartGas.l2Gas - machineState.l2GasLeft,
+          daGas: instrStartGas.daGas - machineState.daGasLeft,
+        };
+        this.tallyInstruction(instrPc, instruction.constructor.name, gasUsed);
```

Comment:
Are you saving the instruction name on purpose so that variants dont matter?

Reply:
Well at first i was using the actual opcode, but then I was getting a lot of `undefined`.... Which I was a bit confused about. Switching to instruction name seemed to resolve that.